src/accelerate/big_modeling.py

```diff
@@ -324,63 +324,73 @@ def dispatch_model(
-
-    if main_device != "cpu":
-        cpu_modules = [name for name, device in device_map.items() if device == "cpu"]
-        if state_dict is None and len(cpu_modules) > 0:
-            state_dict = extract_submodules_state_dict(model.state_dict(), cpu_modules)
-
-    disk_modules = [name for name, device in device_map.items() if device == "disk"]
-    if offload_dir is None and offload_index is None and len(disk_modules) > 0:
-        raise ValueError(
-            "We need an `offload_dir` to dispatch this model according to this `device_map`, the following submodules "
-            f"need to be offloaded: {', '.join(disk_modules)}."
-        )
-    if (
-        len(disk_modules) > 0
-        and offload_index is None
-        and (not os.path.isdir(offload_dir) or not os.path.isfile(os.path.join(offload_dir, "index.json")))
-    ):
-        disk_state_dict = extract_submodules_state_dict(model.state_dict(), disk_modules)
-        offload_state_dict(offload_dir, disk_state_dict)
-
-    execution_device = {
-        name: main_device if device in ["cpu", "disk"] else device for name, device in device_map.items()
-    }
-    execution_device[""] = main_device
-    offloaded_devices = ["disk"] if main_device == "cpu" or main_device == "mps" else ["cpu", "disk"]
-    offload = {name: device in offloaded_devices for name, device in device_map.items()}
-    save_folder = offload_dir if len(disk_modules) > 0 else None
-    if state_dict is not None or save_folder is not None or offload_index is not None:
-        device = main_device if offload_index is not None else None
-        weights_map = OffloadedWeightsLoader(
-            state_dict=state_dict, save_folder=save_folder, index=offload_index, device=device
+            weights_map = None
+
+        tied_params = find_tied_parameters(model)
+        attach_align_device_hook_on_blocks(
+            model,
+            execution_device=execution_device,
+            offload=offload,
+            offload_buffers=offload_buffers,
+            weights_map=weights_map,
+            skip_keys=skip_keys,
+            preload_module_classes=preload_module_classes,
         )
+        # Attaching the hook may break tied weights, so we retie them
+        retie_parameters(model, tied_params)
     else:
-        weights_map = None
-
-    tied_params = find_tied_parameters(model)
-    attach_align_device_hook_on_blocks(
-        model,
-        execution_device=execution_device,
-        offload=offload,
-        offload_buffers=offload_buffers,
-        weights_map=weights_map,
-        skip_keys=skip_keys,
-        preload_module_classes=preload_module_classes,
-    )
-    # Attaching the hook may break tied weights, so we retie them
-    retie_parameters(model, tied_params)
+        device = device_map.values()[0]
+        if device != "disk":
+            model.to(device)
+        else:
+            raise ValueError("You can't offload the whole model to the disk")
```

Comment:
Maybe point the user to the `offload_disk` function in this error message?

Reply:
nice